src/helpers/upgrade/executableHelpers.ts

```diff
@@ -0,0 +1,135 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import { NativefierOptions } from '../../options/model';
+import { getVersionString } from './rceditGet';
+type ExecutableInfo = {
+  arch?: string;
+};
+
+function getExecutableBytes(executablePath: string): Uint8Array {
+  return fs.readFileSync(executablePath);
+}
+
+function getExecutableArch(
```

Comment:
Here and everywhere else: you're going to great lengths to recover information that wasn't serialized to `nativefier.json`.

Your current code looks good and I'm fully able to trust you that it works now, but I'm also a bit wary that future additions will lack such handling, and so `--upgrade` risks becoming subtly broken ("broken" meaning "failing to port one or several existing options").

→ Rather than detecting everything like you do here, what about serializing them in the first place?

... and as I write this, comes to mind the probable reason you did all this: migration cases. If in version 43.0.1 we add serialization of basically everything to `nativefier.json`, it means that the simpler naiver upgrade code I'm suggesting (counting on everything to be in nativefier.json) will choke on apps generated before 43.0.1 😕.

Hm. But let's remind this is a new feature anyway. Maybe supporting upgrading only the >=43.0.1 apps (and failing for earlier apps lacking info in `nativefier.json` with some kind of `Sorry, --upgrade can only upgrade Nativefier >= 43.0.1 apps. You'll have to regenerate manually this time, and --upgrade will work afterwards :)` message would  be reasonable?

Slight inconvenience vs. maintainability tradeoff. Waddayathink?

Reply:
> Here and everywhere else: you're going to great lengths to recover information that wasn't serialized to `nativefier.json`.

Correct. You should see the code I cut because even I thought it was going too far. 🤣 

> Your current code looks good and I'm fully able to trust you that it works now, but I'm also a bit wary that future additions will lack such handling, and so `--upgrade` risks becoming subtly broken ("broken" meaning "failing to port one or several existing options").

That is a risk, yes.

> → Rather than detecting everything like you do here, what about serializing them in the first place?
> 
> ... and as I write this, comes to mind the probable reason you did all this: migration cases. If in version 43.0.1 we add serialization of basically everything to `nativefier.json`, it means that the simpler naiver upgrade code I'm suggesting (counting on everything to be in nativefier.json) will choke on apps generated before 43.0.1 😕.

I walked this same thought process myself 😆 

> Hm. But let's remind this is a new feature anyway. Maybe supporting upgrading only the >=43.0.1 apps (and failing for earlier apps lacking info in `nativefier.json` with some kind of `Sorry, --upgrade can only upgrade Nativefier >= 43.0.1 apps. You'll have to regenerate manually this time, and --upgrade will work afterwards :)` message would be reasonable?
> 
> Slight inconvenience vs. maintainability tradeoff. Waddayathink?

I would propose a middle ground where we begin storing the packager options now, keep this autodetect code for now, and deprecate it in the future.

The reason I say this is this is needed to even get some very basic app information, such as arch, so shipping without this seems a little half-hearted. I try to approach features like this from the end user's perspective, and as an end user I would expect it to figure out things like the arch.

Another idea I've been toying with is splitting out some of this code into a separate package that can extract executable information. If you'd be open, I can finish doing that, and include it as a requirement here. Upside is it cleans up nativefier code, downside is you'd have a new external requirement.

Reply:
@TheCleric 

> I would propose a middle ground where we begin storing the packager options now, keep this autodetect code for now, and deprecate it in the future.

Sounds good!

> Another idea I've been toying with is splitting out some of this code into a separate package that can extract executable information. If you'd be open, I can finish doing that, and include it as a requirement here. Upside is it cleans up nativefier code, downside is you'd have a new external requirement.

Yes to make it as isolated as possible, so that the day we want to deprecate it, it's easy to pull the thread and delete it all. But I wouldn't go as far as pulling into a separate package.

Reply:
Okay, all this code is in the `executableHelper.ts` file for easy removal later.